### PR TITLE
lxd/storage: Only delete volumes from non-pending pools

### PR DESCRIPTION
**Pull request by monstermunchkin**

### CODE REVIEW

**Comment by monstermunchkin** [comment]
I ran into this issue with Ceph version 15.2.7. It doesn't seem to have been a problem on Jenkins but perhaps we're using an older version there.

**Review by tomponline — COMMENTED** [review]
@monstermunchkin to avoid duplication with the default values (I can see at least one other case where "admin" is defined as the default value) how about adding some helper functions on the `ceph` struct, such as d.clusterName(), d.userName() etc, that try to access the config and if not set return the default values, that way you can then use them anywhere you need to and only need to define the default values once.

**Comment by stgraber** [comment]
The title is confusing me, what's the problem exactly?
If the pool is in state pending, then nothing should be allowed on it.

**Comment by monstermunchkin** [comment]
> The title is confusing me, what's the problem exactly?
> If the pool is in state pending, then nothing should be allowed on it.

I see why the title might be confusing, I'll update it as well as the comment in the code.

In our tests we [try to create a container on a pending node](https://github.com/lxc/lxd/blob/master/test/suites/clustering.sh#L465-L468). That isn't allowed and will also fail. However, before it fails, it calls `hasVolume` at some point but is missing `ceph.user.name`, `ceph.osd.pool_name` and `ceph.cluster_name`. The problem is that `rbd --id '' --cluster '' --pool '' info container_bar` will not exit but wait forever. At least it does so with Ceph version 15.2.7. Also, I couldn't find any timeout option for `rbd`.
We therefore need to make sure, we provide at least the default values so that the `rbd` call can fail.

**Comment by stgraber** [comment]
I don't understand why we hit the storage driver at all in this case. We should never be allowed to even retrieve the storage pool is it's in pending state, so this should fail way before the ceph driver.

**Comment by monstermunchkin** [comment]
> I don't understand why we hit the storage driver at all in this case. We should never be allowed to even retrieve the storage pool is it's in pending state, so this should fail way before the ceph driver.

You are right, and [here's the issue](https://github.com/lxc/lxd/blob/master/lxd/instance.go#L155-L172). In this specific case, `GetPoolByInstance()` fails but we still call `inst.Delete()` as part of the revert which then calls `pool.DeleteInstance()` (and the driver code).

**Comment by monstermunchkin** [comment]
@stgraber How about we check whether the pool is pending [here](https://github.com/lxc/lxd/blob/master/lxd/instance/drivers/driver_lxc.go#L3443)?

```go
// ...
			if !isImport && pool.Status != "Pending" {
				// Remove the storage volume, snapshot volumes and database records.
				err = pool.DeleteInstance(d, nil)
				if err != nil {
					return err
				}
			}
// ...
```

**Comment by monstermunchkin** [comment]
Actually, that won't solve it since it would have already called `instanceCreateInternal()`.

**Comment by monstermunchkin** [comment]
@stgraber this fixes the issue.

**Comment by stgraber** [comment]
@tomponline if you're around, what are your thoughts on this?

It feels a bit weird doing this check on Delete(), it feels like we should catch this in instance config validation, not allowing a `pool` property that points to a pending pool.

**Comment by stgraber** [comment]
@tomponline can you think of any reason not to validate any `pool` property in `validateConfig` (disk device) to ensure the pool exists and is in active state?

**Comment by tomponline** [comment]
@stgraber no I can't. It seems like a good idea. I too am not keen on putting pool state conditions on volume delete. However the default value functions @monstermunchkin added also seem like a good idea as we never want the has volume function to hang.

**Comment by stgraber** [comment]
Jenkins isn't looking too happy about this branch.

File: lxd/device/disk.go [hunk 1]
```diff
@@ -178,6 +178,17 @@ func (d *disk) validateConfig(instConf instance.ConfigReader) error {
 	}
 
 	if d.config["pool"] != "" {
+		if d.inst != nil && !d.inst.IsSnapshot() {
+			_, pool, _, err := d.state.Cluster.GetStoragePoolInAnyState(d.config["pool"])
+			if err != nil {
+				return fmt.Errorf("Failed to get storage pool %q: %s", d.config["pool"], err)
```
Comment (tomponline) [inline lxd/device/disk.go:184]: @monstermunchkin should really use `errors.Wrapf(err,...)` here for future reference.

**Pull request merged**